runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java

```diff
@@ -0,0 +1,207 @@
+ *
+ * @param <E>   Type of the entry to be enqueued into the persisted queue
+ */
+@Slf4j
+@CorfuObject
+public class CorfuQueue<E> implements ISMRObject {
+    /** The "main" linked map which contains the primary key-value mappings. */
+    private final Map<UUID, E> mainMap = new LinkedHashMap<>();
+
+    private final OrderedGuidGenerator guidGenerator;
+
+    public CorfuQueue(OrderedGuidGenerator guidGenerator) {
+        this.guidGenerator = guidGenerator;
+    }
+
+    /** Returns the size of the queue at a given point in time. */
+    @Accessor
+    public int size() {
+        return mainMap.size();
+    }
+
+    /**
+     * Appends the specified element at the end of this unbounded queue.
+     * In a distributed system, the order of insertions cannot be guaranteed
+     * unless a transaction is used.
+     * Capacity restrictions and backoffs must be implemented outside this
+     * interface. Consider validating the size of the queue against a high
+     * watermark before enqueue.
+     *
+     * @param e the element to add
+     * @throws NullPointerException if the specified element is null and this
+     *         queue does not permit null elements
+     * @throws IllegalArgumentException if some property of the specified
+     *         element prevents it from being added to this queue
+     * @return Globally ordered and unique ID representing this entry in the persistent queue
+     */
+    public UUID enqueue(E e) {
+        UUID id = guidGenerator.nextUUID();
+        put(id, e);
+        return id;
+    }
+
+    /** CorfuQueueRecord encapsulates each entry enqueued into CorfuQueue with its unique ID.
+     * It is a read-only type returned by the entryList() method.
+     * The ID returned here can be used for both point get()s as well as remove() operations
+     * on this Queue.
+     *
+     * @param <E>
+     */
+    public static class CorfuQueueRecord<E> {
+        /**
+         * This ID represents the entry and its order in the Queue.
+         * This implies that it is unique and comparable with other IDs
+         * returned from CorfuQueue methods with respect to its enqueue order.
+         */
+        @Getter
+        final UUID ID;
+
+        @Getter
+        final E entry;
+
+        CorfuQueueRecord(UUID id, E entry) {
+            this.ID = id;
+            this.entry = entry;
+        }
+    }
+
+    /**
+     * Returns a List of CorfuQueueRecords sorted by the order in which the enqueue materialized.
+     *
+     * {@inheritDoc}
+     *
+     * <p>This function currently does not return a view like the java.util implementation,
+     * and changes to the entryList will *not* be reflected in the map. </p>
+     *
+     * @param maxEntries - Limit the number of entries returned from start of the queue
+     *                   -1 implies all entries
+     * @return List of Entries sorted by their enqueue order
+     */
+    public List<CorfuQueueRecord<E>> entryList(int maxEntries) {
```

Comment:
In my understanding, this API is mostly used in recovery, meaning if the client crashed, lost all its in-memory Guid keys that are not consumed yet, this API helps recover it. Is my understanding correct? If so could you add it to the comment?

Reply:
hi Wenbin, it is not used for recovery alone. This is the main "consumer" of the queue, which returns the entries as an ordered list. I will make this clearer in the Javadoc as you suggest..